provider/core.rb

```diff
@@ -28,6 +28,112 @@ module Provider
     quiet: false
   }.freeze
 
+  # Responsible for generating a file
+  # with a given set of parameters.
+  class FileTemplate
+    include Compile::Core
+    attr_accessor :name
+    attr_accessor :object
+    attr_accessor :product
+    attr_accessor :version
+    attr_accessor :output_folder
+    attr_accessor :product_ns
+    attr_accessor :config
+    attr_accessor :manifest
+    attr_accessor :env
+
+    # Ansible stuff.
+    attr_accessor :example
+
+    # InSpec stuff.
+    attr_accessor :plural
+    attr_accessor :doc_generation
+    attr_accessor :attribute_file_name
+    attr_accessor :privileged
+    attr_accessor :property
+
+    # Terraform stuff.
+    attr_accessor :async
+    attr_accessor :resource_name
+
+    def initialize(options)
+      options.each { |k, v| method("#{k}=").call(v) }
+
+      @env = {
+        pyformat_enabled: options.dig(:env, :pyformat_enabled),
+        goformat_enabled: options.dig(:env, :goformat_enabled),
+        start_time: options.dig(:env, :start_time)
+      }
+    end
+
+    # Given the data object for a file, write that file and verify that it
+    # passes these conditions:
+    #
+    # - The file has not been generated already this run
+    # - The file has an autogen exception or an autogen notice defined
+    #
+    # Once the file's contents are written, set the proper [chmod] mode and
+    # format the file with a language-appropriate formatter.
+    def generate(template, path, provider)
+      folder = File.dirname(path)
+      FileUtils.mkpath folder unless Dir.exist?(folder)
+
+      # If we've modified a file since starting an MM run, it's a reasonable
+      # assumption that it was this run that modified it.
+      if File.exist?(path) && File.mtime(path) > @env[:start_time]
+        raise "#{path} was already modified during this run"
+      end
+
+      # You're looking at some magic here!
+      # This is how variables are made available in templates; we iterate
+      # through each key:value pair in this object, and we set them
+      # in the scope of the provider.
+      #
+      # The templates get access to everything in the provider +
+      # all of the variables in this object.
+      ctx = provider.provider_binding
+      instance_variables.each do |name|
+        ctx.local_variable_set(name[1..-1], instance_variable_get(name))
+      end
+
+      # This variable looks unused, but is used in ansible/resource.erb
```

Comment:
I don't think this comment makes sense post-refactor

Reply:
changed.